File: backend/app/services/scrapers/validators/url_validator.py
from app.schemas.scraped_job import ScrapedJob
from app.services.scrapers.validators.scraped_job_validator import ScrapedJobValidator


class URLValidator(ScrapedJobValidator):
    def _do_validate(self, job: ScrapedJob) -> bool:
        try:
            # Must exist and be a valid HTTP web link
            if not job.url or not job.url.startswith("http"):
                return False
                
            # Filter out internal ATS test links if any leak through
            if "test" in job.url.lower() or "demo" in job.url.lower():
                return False
                
        except Exception as e:
            print(f"[URLValidator] Error processing URL: {e}")
            return False

        return self.pass_to_next_validator(job)

File: job-scraper-service/internal/validators/url_validator.go
package validator

import (
	"strings"

	"job-scraper-service/internal/models"
)

type URLValidator struct {
	BaseValidator
}

func NewURLValidator() *URLValidator {
	return &URLValidator{
		BaseValidator: NewBaseValidator(),
	}
}

func (v *URLValidator) Validate(job models.ScrapedJob) bool {
	// Must exist and be a valid HTTP web link
	if job.URL == "" || !strings.HasPrefix(job.URL, "http") {
		v.RecordStat(job.Platform, "fail")
		return false
	}

	// Filter out internal ATS test links if any leak through
	lowerURL := strings.ToLower(job.URL)
	if strings.Contains(lowerURL, "test") || strings.Contains(lowerURL, "demo") {
		v.RecordStat(job.Platform, "fail")
		return false
	}

	v.RecordStat(job.Platform, "pass")
	return v.PassToNext(job)
}
